internal/lib/sandbox/namespaces.go

```diff
@@ -114,72 +78,30 @@ func (m *ManagedNamespace) Path() string {
 	return m.nsPath
 }
 
-// CreateManagedNamespaces calls pinnsPath on all the managed namespaces for the sandbox.
-// It returns a slice of ManagedNamespaces it created.
-func (s *Sandbox) CreateManagedNamespaces(managedNamespaces []NSType, idMappings *idtools.IDMappings, sysctls map[string]string, cfg *config.Config) ([]*ManagedNamespace, error) {
-	return s.CreateNamespacesWithFunc(managedNamespaces, idMappings, sysctls, cfg, pinNamespaces)
-}
-
-type namespacePinner func([]NSType, *config.Config, *idtools.IDMappings, map[string]string) ([]NamespaceIface, error)
-
-// CreateManagedNamespacesWithFunc is mainly added for testing purposes. There's no point in actually calling the pinns binary
-// in unit tests, so this function allows the actual pin func to be abstracted out. Every other caller should use CreateManagedNamespaces
-func (s *Sandbox) CreateNamespacesWithFunc(managedNamespaces []NSType, idMappings *idtools.IDMappings, sysctls map[string]string, cfg *config.Config, pinFunc namespacePinner) (mns []*ManagedNamespace, retErr error) {
-	typesAndPaths := make([]*ManagedNamespace, 0, 4)
-	if len(managedNamespaces) == 0 {
-		return typesAndPaths, nil
-	}
-
-	namespaces, err := pinFunc(managedNamespaces, cfg, idMappings, sysctls)
-	if err != nil {
-		return nil, err
+func (s *Sandbox) AddManagedNamespaces(namespaces []nsmgr.NamespaceIface) {
+	// if the namespace structure wasn't initialized, we have nothing to do here
+	if namespaces == nil {
+		return
 	}
-
-	for _, namespace := range namespaces {
-		namespaceIface := namespace.Initialize()
-		defer func() {
-			if retErr != nil {
-				if err1 := namespaceIface.Remove(); err1 != nil {
-					logrus.Warnf("removing namespace interface returned: %v", err1)
-				}
-			}
-		}()
-
-		switch namespace.Type() {
-		case NETNS:
-			s.netns = namespaceIface
-			typesAndPaths = append(typesAndPaths, &ManagedNamespace{
-				nsType: NETNS,
-				nsPath: namespace.Path(),
-			})
-		case IPCNS:
-			s.ipcns = namespaceIface
-			typesAndPaths = append(typesAndPaths, &ManagedNamespace{
-				nsType: IPCNS,
-				nsPath: namespace.Path(),
-			})
-		case UTSNS:
-			s.utsns = namespaceIface
-			typesAndPaths = append(typesAndPaths, &ManagedNamespace{
-				nsType: UTSNS,
-				nsPath: namespace.Path(),
-			})
-		case USERNS:
-			if idMappings != nil {
-				s.userns = namespaceIface
-				typesAndPaths = append(typesAndPaths, &ManagedNamespace{
-					nsType: USERNS,
-					nsPath: namespace.Path(),
-				})
-			}
+	for _, ns := range namespaces {
+		// skip any nil entries
+		if ns == nil {
+			continue
+		}
+		switch ns.Type() {
+		case nsmgr.IPCNS:
+			s.ipcns = ns
+		case nsmgr.UTSNS:
+			s.utsns = ns
+		case nsmgr.NETNS:
+			s.netns = ns
+		case nsmgr.USERNS:
+			s.userns = ns
+		// this should never happen
 		default:
-			// This should never happen
-			err = errors.New("Invalid namespace type")
-			return typesAndPaths, err
+			continue
```

Comment:
Why are unknown entries skipped? If "this should never happen" and there's no way / need to return an error, the code can `panic()` here.

Reply:
fixed!